exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,542 @@
+                        JValue::from(node_handle),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+/// Artifact identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct JavaArtifactId(String);
+
+impl fmt::Display for JavaArtifactId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceStateHashes")]
+struct ServiceStateHashes {
+    instance_id: u32,
+    state_hashes: Vec<Vec<u8>>,
+}
+
+impl From<&ServiceStateHashes> for (InstanceId, Vec<Hash>) {
+    fn from(value: &ServiceStateHashes) -> Self {
+        let hashes: Vec<Hash> = value
+            .state_hashes
+            .iter()
+            .map(|bytes| to_hash(bytes))
+            .collect();
+        (value.instance_id, hashes)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceRuntimeStateHashes")]
+struct ServiceRuntimeStateHashes {
+    runtime_state_hashes: Vec<Vec<u8>>,
+    service_state_hashes: Vec<ServiceStateHashes>,
+}
+
+impl ServiceRuntimeStateHashes {
+    fn runtime(&self) -> Vec<Hash> {
+        self.runtime_state_hashes
+            .iter()
+            .map(|bytes| to_hash(bytes))
+            .collect()
+    }
+
+    fn instances(&self) -> Vec<(InstanceId, Vec<Hash>)> {
+        self.service_state_hashes
+            .iter()
+            .map(|service| service.into())
+            .collect()
+    }
+}
+
+impl From<ServiceRuntimeStateHashes> for StateHashAggregator {
+    fn from(value: ServiceRuntimeStateHashes) -> Self {
+        StateHashAggregator {
+            runtime: value.runtime(),
+            instances: value.instances(),
+        }
+    }
+}
+
+fn to_hash(bytes: &Vec<u8>) -> Hash {
+    Hash::from_bytes(bytes.into()).unwrap()
+}
+
+type ExceptionHandler = Fn(&JNIEnv, JObject) -> ExecutionError;
+struct ExceptionHandlers;
+
+impl ExceptionHandlers {
+    const DEFAULT: &'static ExceptionHandler = &|env, exception| {
+        assert!(!exception.is_null(), "No exception thrown.");
+        let message = describe_java_exception(env, exception);
+        (Error::JavaException, message).into()
+    };
+
+    const TX_EXECUTION: &'static ExceptionHandler = &|env, exception| {
+        assert!(!exception.is_null(), "No exception thrown.");
+        let code = unwrap_jni(Self::get_tx_error_code(env, exception)) as u8;
+        let msg = unwrap_jni(get_exception_message(env, exception))
+            .unwrap_or(String::new());
+        ExecutionError::new(ErrorKind::service(code), msg)
+    };
+
+    const SERVICE_LOADING: &'static ExceptionHandler = &|env, exception| {
```

Comment:
Do we need to handle service loading exception specially (it is converted to JavaException just as the default)? Is there any benefit in defining more specific runtime errors or will JavaException do in this case?